### PR TITLE
Support read and write ByteBuffer in ServletInputStream and ServletOutputStream

### LINKED ISSUE #26500: Servlet 6.1: Add ByteBuffer support to ServletInputStream and ServletOutputStream.
`https://github.com/jakartaee/servlet/issues/484`

### PR DESCRIPTION
**Pull request by pmd1nh**
fixes https://github.com/OpenLiberty/open-liberty/issues/26500

### CODE REVIEW

**Comment by pmd1nh** [comment]
PR for the ServletOutputStream write(ByteBuffer)

**Comment by pmd1nh** [comment]
#build
        #spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat,io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.6.0_fat_toleration.lWAS,com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat,io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.6.0_fat_toleration.lWAS,com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build

**Comment by pmd1nh** [comment]
#build

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat
#libby

**Comment by posmith** [comment]
Support message review completed.

**Comment by pmd1nh** [comment]
1 failed but unrelated..due to timeout

```
io.openliberty.springboot.fat30_fat FAT tests 1 0 1
 com.ibm.ws.springboot.support.fat.FATSuite 1 0 1 0 s 1
  testBlockingIOServlet60 1 0 1 0 s 1
```

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 1]
```diff
@@ -1633,3 +1633,11 @@ redirect.location.cannot.be.null.useraction=Update the redirect location by spec
 sendRedirect.hyper.text.url=<html><body><p>Redirecting to <a href="{0}"></a></p></body></html>
 
 query.or.post.is.null=Request query string or post data is null.
+
+read.write.bytebuffer.null=CWWWC0009E: An attempt to read from an InputStream or write to an OutputStream failed because a variable argument ByteBuffer was null.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]:
```suggestion
read.write.bytebuffer.null=CWWWC0009E: An attempt to read from an InputStream class or write to an OutputStream class failed because a variable argument ByteBuffer class was null.
```

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 2]
```diff
@@ -1633,3 +1633,11 @@ redirect.location.cannot.be.null.useraction=Update the redirect location by spec
...
+read.write.bytebuffer.null.explanation=The variable argument ByteBuffer passed to the read or write API must not be null.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]:
```suggestion
read.write.bytebuffer.null.explanation=The variable argument ByteBuffer class that is passed to the read or write API must not be null.
```

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 3]
```diff
@@ -1633,3 +1633,11 @@ redirect.location.cannot.be.null.useraction=Update the redirect location by spec
...
+read.write.bytebuffer.null.useraction=Modify the application code according to the error explanation.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: can this be more specific? How do make a ByteBuffer not be null?

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 4]
```diff
@@ -1633,3 +1633,11 @@ redirect.location.cannot.be.null.useraction=Update the redirect location by spec
...
+
+read.write.failed.isReady.false=CWWWC0010E: An attempt to read from an InputStream or write to an OutputStream failed because isReady API returns false.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]:
```suggestion
read.write.failed.isReady.false=CWWWC0010E: An attempt to read from an InputStream class or write to an OutputStream class failed because the isReady API returned false.
```

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 5]
```diff
@@ -1633,3 +1633,11 @@ redirect.location.cannot.be.null.useraction=Update the redirect location by spec
...
+read.write.failed.isReady.false.explanation= If isReady returns false it is illegal to call a read or write method.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]:
```suggestion
read.write.failed.isReady.false.explanation= If the isReady API returns false, you cannot call a read or write method.
```

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 6]
```diff
@@ -1633,3 +1633,11 @@ redirect.location.cannot.be.null.useraction=Update the redirect location by spec
...
+read.write.failed.isReady.false.useraction= Modify the application code according to the error explanation.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: here also- can we be more specific?

**Review by dmuelle — CHANGES_REQUESTED** [review]
ID review

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops (hunk 6 above)
Comment (pmd1nh) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: @dmuelle I don't know how to explain or be more specific about this one.  The API doc for isReady()

https://jakarta.ee/specifications/servlet/6.0/apidocs/jakarta.servlet/jakarta/servlet/servletinputstream#isReady()

this flag is dynamically changed at runtime depending on if the stream is busy writing or reading when another async resource is trying to query for isReady().

Comment (dmuelle) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: In this case- would the operation just be retried later to see if isReady is still busy? How might the user adjust the application code to avoid the error? We are telling them to edit the app code according to the error explanation, but the explanation doesnt say anything about how to edit the code, how persistent the issue might be or what is causing it.

Comment (pmd1nh) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: @dmuelle  How about:

read.write.bytebuffer.null.useraction=Modify the application to create a non-null ByteBuffer before using in an InputStream or OutputStream.


read.write.failed.isReady.false.useraction= Modify the application logic to make sure the stream is ready for an InputStream or OutputStream operation.

File: dev/io.openliberty.webcontainer.servlet.6.1.internal/src/io/openliberty/webcontainer61/osgi/response/WCOutputStream61.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        byte[] b = new byte[buffer.remaining()];
+        buffer.get(b);
+
+        super.write(b);
+
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.exit(tc, "write(ByteBuffer)");
+        }
+    }
+}
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.1.internal/src/io/openliberty/webcontainer61/osgi/response/WCOutputStream61.java]: new line

File: dev/io.openliberty.webcontainer.servlet.6.1.internal_fat/publish/servers/servlet61_ReadWriteByteBuffer/bootstrap.properties [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License 2.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-2.0/
+# 
+# SPDX-License-Identifier: EPL-2.0
+###############################################################################
+bootstrap.include=../testports.properties
+osgi.console=7777
+com.ibm.ws.logging.trace.specification=*=info:io.openliberty.webcontainer*=all:com.ibm.ws.webcontainer*=all:com.ibm.wsspi.webcontainer*=all:HTTPChannel=all:GenericBNF=all:HTTPDispatcher=all
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.1.internal_fat/publish/servers/servlet61_ReadWriteByteBuffer/bootstrap.properties]: Lets remove all the tracing before merging if it is not necessary for the test.

File: dev/io.openliberty.webcontainer.servlet.6.1.internal_fat/test-applications/ReadWriteByteBuffer.war/src/readwritebytebuffer/servlets/ReadWriteByteBuffer.java [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    private void LOGENTER(String method) {
+        LOG(">>>>>>>>>>>>>>>> TESTING [" + method + "] ENTER >>>>>>>>>>>>>>>>");
+    }
+
+    private void LOGEXIT(String method) {
+        LOG("<<<<<<<<<<<<<<<<<< TESTING [" + method + "] EXIT <<<<<<<<<<<<<<<<<<");
+    }
+}
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.1.internal_fat/test-applications/ReadWriteByteBuffer.war/src/readwritebytebuffer/servlets/ReadWriteByteBuffer.java]: new line

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops (hunk 6 above)
Comment (dmuelle) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: Looks good, I'd suggest a slight edit to the first one

> ead.write.bytebuffer.null.useraction=Modify the application to create a non-null ByteBuffer **before using** in an InputStream or OutputStream.


What is being used in an I/O stream? the application or the ByteBuffer?

Comment (pmd1nh) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: The application needs to create a ByteBuffer which is passed into the Input/OutputStream. @dmuelle

**Comment by pmd1nh** [comment]
#libby

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 10]
```diff
@@ -1635,3 +1635,11 @@ sendRedirect.hyper.text.url=<html><body><p>Redirecting to <a href="{0}"></a></p>
 query.or.post.is.null=CWWWC0009E: The HTTP request query string or HTTP POST request data is null.
 query.or.post.is.null.explanation=The application tried to parse a null HTTP request query string or a null HTTP POST request body.
 query.or.post.is.null.useraction=The client must send a non-null HTTP request query string or a non-null HTTP POST request body.
+
+read.write.bytebuffer.null=CWWWC0010E: An attempt to read from an InputStream or write to an OutputStream failed because a variable argument ByteBuffer was null.
+read.write.bytebuffer.null.explanation=The variable argument ByteBuffer passed to the read or write API must not be null.
+read.write.bytebuffer.null.useraction=Modify the application code according to the error explanation.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]:
```suggestion
read.write.bytebuffer.null.useraction=Modify the application code to create a non-null ByteBuffer instance to pass to the InputStream or OutputStream operation.
```

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 11]
```diff
@@ -1635,3 +1635,11 @@ sendRedirect.hyper.text.url=<html><body><p>Redirecting to <a href="{0}"></a></p>
...
+
+read.write.failed.isReady.false=CWWWC0011E: An attempt to read from an InputStream or write to an OutputStream failed because isReady API returns false.
+read.write.failed.isReady.false.explanation= If isReady returns false it is illegal to call a read or write method.
+read.write.failed.isReady.false.useraction= Modify the application code according to the error explanation.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]:
```suggestion
read.write.failed.isReady.false.useraction= Modify the application logic to make sure that the stream is ready for an InputStream or OutputStream operation.
```

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops (hunk 6 above)
Comment (dmuelle) [reply dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]: ok thanks- left inline suggestions for both these

**Comment by pmd1nh** [comment]
@dmuelle Please give it a final review.  Thanks!

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat,io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.6.0_fat_toleration.lWAS,com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 12]
```diff
@@ -1635,3 +1635,11 @@ sendRedirect.hyper.text.url=<html><body><p>Redirecting to <a href="{0}"></a></p>
 query.or.post.is.null=CWWWC0009E: The HTTP request query string or HTTP POST request data is null.
 query.or.post.is.null.explanation=The application tried to parse a null HTTP request query string or a null HTTP POST request body.
 query.or.post.is.null.useraction=The client must send a non-null HTTP request query string or a non-null HTTP POST request body.
+
+read.write.bytebuffer.null=CWWWC0010E: An attempt to read from an InputStream or write to an OutputStream failed because a variable argument ByteBuffer was null.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]:
```suggestion
read.write.bytebuffer.null=CWWWC0010E: An attempt to read from an InputStream operation or write to an OutputStream operation failed because a variable argument ByteBuffer instance was null.
```

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 13]
```diff
@@ -1635,3 +1635,11 @@ sendRedirect.hyper.text.url=<html><body><p>Redirecting to <a href="{0}"></a></p>
...
+read.write.bytebuffer.null.explanation=The variable argument ByteBuffer passed to the read or write API must not be null.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]:
```suggestion
read.write.bytebuffer.null.explanation=The variable argument ByteBuffer instance that is passed to the read or write API must not be null.
```

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 14]
```diff
@@ -1635,3 +1635,11 @@ sendRedirect.hyper.text.url=<html><body><p>Redirecting to <a href="{0}"></a></p>
...
+read.write.bytebuffer.null.useraction=Modify the application code to create a non-null ByteBuffer instance to pass to the InputStream or OutputStream operation.
+
+read.write.failed.isReady.false=CWWWC0011E: An attempt to read from an InputStream or write to an OutputStream failed because isReady API returns false.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]:
```suggestion
read.write.failed.isReady.false=CWWWC0011E: An attempt to read from an InputStream operation or write to an OutputStream operation failed because the isReady API returns false.
```

File: dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops [hunk 15]
```diff
@@ -1635,3 +1635,11 @@ sendRedirect.hyper.text.url=<html><body><p>Redirecting to <a href="{0}"></a></p>
...
+read.write.failed.isReady.false.explanation=If isReady returns false it is illegal to call a read or write method.
```
Comment (dmuelle) [inline dev/com.ibm.ws.webcontainer/resources/com/ibm/ws/webcontainer/resources/Messages.nlsprops]:
```suggestion
read.write.failed.isReady.false.explanation=If the isReady API returns false, you cannot call a read or write method.
```

**Review by dmuelle — COMMENTED** [review]
ID review- @pmd1nh - the updates we discussed look good- just a few more minor suggestions per IBM style.

**Comment by pmd1nh** [comment]
@dmuelle updated to use IBM style.  Thanks!

**Review by dmuelle — DISMISSED** [review]
LGTM, thanks!

**Comment by pmd1nh** [comment]
@pnicolucci all are updated.

File: dev/com.ibm.ws.webcontainer.servlet.3.1/src/com/ibm/ws/webcontainer31/osgi/response/WCOutputStream31.java [hunk 16]
```diff
@@ -226,8 +229,12 @@ public void write(byte[] b, int start, int len) throws IOException
      * @param len
      * @throws IOException
      */
-    private void write_NonBlocking(byte[] value, int start, int len) throws IOException
+    protected void write_NonBlocking(byte[] value, int start, int len) throws IOException
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer.servlet.3.1/src/com/ibm/ws/webcontainer31/osgi/response/WCOutputStream31.java]: One last thing, why was this changed from private to protected?

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat,io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.6.0_fat_toleration.lWAS,com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

File: dev/io.openliberty.webcontainer.servlet.6.1.internal_fat/fat/src/io/openliberty/webcontainer/servlet61/fat/tests/Servlet61ReadWriteByteBufferTest.java [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+    @Server("servlet61_ReadWriteByteBuffer")
+    public static LibertyServer server;
+
+    @BeforeClass
+    public static void setUp() throws Exception {
+        LOG.info("Setup : servlet61_ReadWriteByteBuffer.");
+        ShrinkHelper.defaultDropinApp(server, TEST_APP_NAME + ".war", "readwritebytebuffer.servlets");
+
+        ArrayList<String> expectedErrors = new ArrayList<String>();
+        expectedErrors.add("CWWWC0009E:.*"); //Write Null ByteBuffer
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.1.internal_fat/fat/src/io/openliberty/webcontainer/servlet61/fat/tests/Servlet61ReadWriteByteBufferTest.java]: Is this correct? We're adding `CWWWC0010E` and `CWWWC0011E` in the product code.

File: dev/io.openliberty.webcontainer.servlet.6.1.internal_fat/test-applications/ReadWriteByteBuffer.war/src/readwritebytebuffer/servlets/ReadWriteByteBuffer.java [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+                LOG("ServletOutputStream.write(ByteBuffer) with NULL ByteBuffer");
+                sos.write(buffer);
+                LOG("write(ByteBuffer) DONE");
+            }
+        }
+        catch (Exception e) {
+            LOG("Exception ["+ (message = e.getMessage()) +"]");
+
+            if (message.contains("CWWWC0009E")) {
+                responseSB.append(method + " found the message with CWWWC0009E. PASS");
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.1.internal_fat/test-applications/ReadWriteByteBuffer.war/src/readwritebytebuffer/servlets/ReadWriteByteBuffer.java]: Does this test pass right now?

This isn't a new message added by this PR.

> query.or.post.is.null=CWWWC0009E: The HTTP request query string or HTTP POST request data is null.
> query.or.post.is.null.explanation=The application tried to parse a null HTTP request query string or a null HTTP POST request body.
> query.or.post.is.null.useraction=The client must send a non-null HTTP request query string or a non-null HTTP POST request body.
> 
> read.write.bytebuffer.null=CWWWC0010E: An attempt to read from an InputStream operation or write to an OutputStream operation failed because a variable argument ByteBuffer instance was null.
> read.write.bytebuffer.null.explanation=The variable argument ByteBuffer instance that is passed to the read or write API must not be null.
> read.write.bytebuffer.null.useraction=Modify the application code to create a non-null ByteBuffer instance to pass to the InputStream or OutputStream operation.
> 
> read.write.failed.isReady.false=CWWWC0011E: An attempt to read from an InputStream operation or write to an OutputStream operation failed because the isReady API returns false.
> read.write.failed.isReady.false.explanation=If the isReady API returns false, you cannot call a read or write method.
> read.write.failed.isReady.false.useraction=Modify the application logic to make sure that the stream is ready for an InputStream or OutputStream operation.

`CWWWC0010E` and `CWWWC0011E` are for this PR, are we testing these new messages?

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat,io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.6.0_fat_toleration.lWAS,com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
#build
#spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.1.internal_fat

**Review by pnicolucci — APPROVED** [review]
PR looks good to me! Make sure you get a good personal build and validate the new tests pass.

**Comment by pmd1nh** [comment]
#libby

**Comment by pmd1nh** [comment]
```
io.openliberty.webcontainer.servlet61.fat.tests.Servlet61ReadWriteByteBufferTest 3 0 0 0 s 8
 test_Stream_ReadWriteByteBuffer 1 0 0 0 s 1
 test_Stream_WriteNullByteBuffer 1 0 0 0 s 2
 test_Stream_ReadNullByteBuffer 1 0 0 0 s 3
```
